### DISCUSSION TITLE
Do you guys want the Python nn and optimizer modules written in C++?

### DISCUSSION THREAD

**Discussion by cemlyn007** [ml-explore/mlx#774] [Q&A] [answered]
Am wondering if you guys would be interested in contributions of implementations of layers and optimizers in C++?

Reason is otherwise people who want to use layers or optimizers will either need to stick with Python (as you guys have implemented it in Python) or they have to implement it themselves which might hurt the barrier to entry.

**Comment by Blakezulla** [comment]
May fix the interference

**Comment by awni** [comment, marked as answer]
It's something we've discussed adding in the past, but haven't quite found enough of a reason to do it. At the moment we're not planning to add that level in C++ until there is enough demand for it (as it adds another surface for us to maintain).

FWIW we also have MLX Swift which has all the high-level layers + otpimizers: https://github.com/ml-explore/mlx-swift/. I'm not sure what your goals are but maybe that would be a good alternative.

However, if this is something super useful for you, I would encourage you to start a separate project which builds some of those layers out on top of MLX in C++. That would make it 1. more concrete so we can see how it looks and if it makes sense to include and 2. see if it is used enough to merit pulling into core.

**Reply by cemlyn007** [reply]
100% makes sense, for now I'll try porting some of the layers and optimizers that I need for my [project](https://github.com/cemlyn007/a3c). I'm not a C++ expert but I'll be giving it a crack over the course of this week. Depending on how it goes I might consider creating a separate project for hosting some implemented layers and optimizers! Worse case my project can be a little example for others.

Thank you and hope you have a good weekend!

**Comment by Blakezulla** [comment]
Thank you for the info
